### ISSUE TITLE
MPS RNG state fails to progress immediately after fork_rng

### ISSUE THREAD

**Issue by ringohoffman** [pytorch/pytorch#124833]
### 🐛 Describe the bug

You have to both first progress the RNG state and then fork. If you comment out either of the first 2 steps, it works as expected. ALSO... this only happens the very first time. If you put this in a for loop, copy-paste this code 2x, or put this in jupyter cell and run it twice, the bug only occurs on the first execution.

```python
import torch

torch.randn(1, device="mps")  # sample

with torch.random.fork_rng(device_type="mps"):  # fork rng
    ...

before_state = torch.mps.get_rng_state()
torch.randn(1, device="mps")  # sample again
after_state = torch.mps.get_rng_state()

print("RNG state progressed:", not torch.allclose(before_state, after_state))  # expected: True, actual: False
```

The equivalent code using cuda works as expected.

### Versions

```console
$ python collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 14.2.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.28.3
Libc version: N/A

Python version: 3.8.18 (default, Sep 11 2023, 08:17:16)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.2.1-arm64-arm-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] optree==0.9.1
[pip3] pytorch-sphinx-theme==0.0.24
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] optree                    0.9.1                    pypi_0    pypi
[conda] pytorch-sphinx-theme      0.0.24                    dev_0    <develop>
```

cc @ezyang @gchanan @zou3519 @kadeng @pbelevich @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by skotapati** [comment]
Hi @ringohoffman thanks for filing this issue. I was unable to reproduce the issue on the ToT, could you try again with the latest nightly build? If the error is still observed, could you provide information regarding your hardware and OS config? Thank you

**Comment by bghira** [comment]
is it this?

```py
    @run_and_save_rng_state.py_impl(DispatchKey.MPS)
    def impl_mps(op, *args, **kwargs):
        return torch.mps.get_rng_state(), op(*args, **kwargs)

    @run_and_save_rng_state.py_impl(DispatchKey.BackendSelect)
    def impl_backend_select(op, *args, **kwargs):
        impl_map = {"cuda": impl_cuda, "cpu": impl_cpu, "mps": impl_mps}
        device = get_device(args, kwargs)
        assert device in impl_map, f"Backend not supported for {device}"
        impl = impl_map[device]
        return impl(op, *args, **kwargs)
```

the implementation is simply missing on mps in the latest nightly.

**Comment by kulinseth** [comment]
@ringohoffman , please open this issue if you think issue persists on latest nightly..

**Comment by bghira** [comment]
yes, it persists on 0511 build.

**Comment by bghira** [comment]
```
 % python repo.py 
RNG state progressed: False
```

```py
import torch

torch.randn(1, device="mps")  # sample

with torch.random.fork_rng(device_type="mps"):  # fork rng
    ...

before_state = torch.mps.get_rng_state()
torch.randn(1, device="mps")  # sample again

after_state = torch.mps.get_rng_state()

print("RNG state progressed:", not torch.allclose(before_state, after_state))  # expected: True, actual: False
```

```
2.4.0.dev20240511
```

**Comment by ringohoffman** [comment]
I think it's fixed as of 2.5.

**Comment by bghira** [comment]
unfortunately 2.5 is not usable on MPS due to memory leaks
